puttyconfig: Switch to string-based Validity format and deprecate MatchHosts

webvictim opened the pull request:
At some point during the development of PuTTY 0.78 which added support for SSH user and host certificates, the author switched from storing the list of hostnames to trust for a given public key in a REG_MULTI_SZ called `MatchHosts`, to storing them in a REG_SZ called `Validity`. 

`tsh puttyconfig` was written to use the `MatchHosts` format, PuTTY still contains code to process that format and tsh only needs to work with single hostnames or wildcards, so there was no sense in delaying the PR further by switching to the more-complicated string-based `Validity` format instead which involves boolean logic and a tokenizer/lexer etc.

However, the author of WinSCP is now adding support for importing saved sessions from and automatically using the same SSH host CAs as configured in PuTTY. WinSCP only supports the newer `Validity` format, so it's time for this PR which switches `tsh puttyconfig` to the same.

If a `MatchHosts` value is set in the registry key, `tsh puttyconfig` will automatically migrate any entries to the `Validity` format, save the key to the registry and then delete the `MatchHosts` value if this was successful.

~Additionally, Teleport ignores any values from the `Validity` key which are inside parentheses and passes them through verbatim to avoid needing to implement a lexer of its own. I suspect it's unlikely that users will manually tinker with the Validity values in keys added by `tsh puttyconfig`, but I wanted to do to try and account for this anyway.~

`tsh puttyconfig` will error if the `Validity` key does not match the managed `<hostname> || <hostname> || ...` format that it uses. There is now a troubleshooting section added to the docs to help people fix these errors if they encounter them.

References:
- PuTTY `Validity` format: https://the.earth.li/~sgtatham/putty/0.79/htmldoc/Chapter4.html#config-ssh-cert-valid-expr
- PuTTY lexer code:  https://git.tartarus.org/?p=simon/putty.git;a=blob;f=utils/cert-expr.c

webvictim committed:
puttyconfig: Switch to string-based Validity format and deprecate MatchHosts

webvictim commented on lib/puttyhosts/puttyhosts.go:
```diff
@@ -238,3 +238,42 @@ func FormatHostCAPublicKeysForRegistry(hostCAPublicKeys map[string][]string, hos
+	// If any values are encountered inside parentheses (), ignore them and pass them through as-is.
+	// This is done to avoid implementing a full lexer as in https://git.tartarus.org/?p=simon/putty.git;a=blob;f=utils/cert-expr.c
+	// because we only care about the individual host elements inside.
+	for i := 0; i < len(input); i++ {
+		char := rune(input[i])
+
+		if char == '(' {
+			stack = append(stack, char)
+			currentToken += string(char)
+		} else if char == ')' {
+			if len(stack) > 0 {
+				stack = stack[:len(stack)-1]
+				currentToken += string(char)
+			}
+		} else if i <= len(input)-2 && input[i:i+2] == "||" && len(stack) == 0 {
+			output = append(output, strings.TrimSpace(currentToken))
+			currentToken = ""
+			i += 1 // skip past the other character in the token
+		} else {
+			currentToken += string(char)
+		}
+	}
```
This works well but seems a little gross, so any feedback on a tidier/more efficient way to implement it is gratefully received.

webvictim commented on lib/puttyhosts/puttyhosts.go:
```diff
@@ -238,3 +238,42 @@ func FormatHostCAPublicKeysForRegistry(hostCAPublicKeys map[string][]string, hos
+	// If any values are encountered inside parentheses (), ignore them and pass them through as-is.
+	// This is done to avoid implementing a full lexer as in https://git.tartarus.org/?p=simon/putty.git;a=blob;f=utils/cert-expr.c
+	// because we only care about the individual host elements inside.
+	for i := 0; i < len(input); i++ {
+		char := rune(input[i])
+
+		if char == '(' {
+			stack = append(stack, char)
+			currentToken += string(char)
+		} else if char == ')' {
+			if len(stack) > 0 {
+				stack = stack[:len(stack)-1]
+				currentToken += string(char)
+			}
+		} else if i <= len(input)-2 && input[i:i+2] == "||" && len(stack) == 0 {
+			output = append(output, strings.TrimSpace(currentToken))
+			currentToken = ""
+			i += 1 // skip past the other character in the token
+		} else {
+			currentToken += string(char)
+		}
+	}
+
+	if len(currentToken) > 0 {
+		output = append(output, strings.TrimSpace(currentToken))
+	}
```
This works well but seems a little gross, so any feedback on a tidier/more efficient way to implement it is gratefully received.

zmb3 replied:
@nklaassen is very good at parsing/validating this type of expression language. Worth checking with him if there's anything we might be able to reuse or at least follow as an example.

nklaassen replied:
Lexer code makes me squirm, it's so hard to get right. As is, this seems like it could behave badly (generate invalid expressions) when there are other operators (`&&` or `!`) or unmatched parens present.

I have really only been building stuff with gravitational/predicate which uses go/parser under the hood, which I don't think would handle this syntax. Is there any way it's possible to avoid parsing this at all? Could we make it write-only, completely regenerated each time we need to?

If we can't avoid reading it completely, could we only touch it if we can make sure it strictly follows the pattern `a || b || c` by:
  * first splitting on whitespace
  * ensure there's an odd number of tokens
  * ensure every odd indexed token is exactly `||`
  * ensure there is no `&&`, `!` or parens anywhere

If this all passes I think we could safely add an `|| hostname` at the end. I'm thinking/hoping it's not necessary to support parsing expressions after users have been messing with them too much.

webvictim replied:
The test cases work OK with `&&` and matched parens, but I haven't tested `!`. I did also test this earlier with mismatched parens, as well as using `|` instead of `||` and you're right, it does break in those cases.

Unfortunately there's no way to make this both functional _and_ write-only - each time a new host is added, it's essential that we preserve all the hosts which were previously added (hopefully always by `tsh puttyconfig`) and then append the new host to the end. The only alternative would be to try to re-read and parse every saved Teleport session from PuTTY's saved session list, which would itself introduce considerable instability.

Your feedback has given me some good ideas for how to make this more resilient, though. I think there are two good options:

1) Add additional validation so that we don't operate on `Validity` strings which are obviously broken. PuTTY is going to reject a non-parsing string when attempting to connect anyway, so there's no real change in functionality. Some additional checks we could do:
- check that the count of `(` is equal to the count of `)`, and error out if not
- check that every instance of `|` is immediately followed by `|`, and error out if not
- check that every instance of `&` is immediately followed by `&`, and error out if not
- error out if the string contains `!` or `:`, as Teleport has no need to make use of this functionality and we don't need to support others introducing it.

2) Reimplement this exactly as you described above and just error out if the string doesn't match that `<value> || <value> || ...` pattern. I don't see any need for us to actually support people modifying the validity outside of Teleport, and if they do it's fair to say they should expect that things will break.

The only real downside with both of these approaches is that it'll make us more likely to have to show errors to the end user, but I can add an additional steps to the troubleshooting section of the documentation detailing how to manually clear a broken Validity list. I could also potentially implement a `tsh puttyconfig --clear-host-cas` command for less technical users which purges all Teleport-managed Host CAs from the registry and starts afresh, although I'd prefer not to creep the scope of this PR in the hope that I can merge this for 14.0.1, or 14.0.2 at the latest.

webvictim replied:
This has been reimplemented using option 2, and I have added a troubleshooting step to the docs with instructions on how to fix any resulting validity errors.

webvictim committed:
Switch to more restrictive, reliable parsing

webvictim committed:
Add validity string errors to docs

webvictim committed:
Remove invalid test case

webvictim committed:
Add test case

webvictim committed:
Remove any spaces from user-provided input and use sanitized hostname

nklaassen commented on lib/puttyhosts/puttyhosts_test.go:
```diff
@@ -305,3 +309,132 @@ func TestNaivelyValidateHostname(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.input, func(t *testing.T) {
+			testResult, err := CheckAndSplitValidityKey(tt.input, "TeleportHostCA-testcluster.example.com")
+			if err != nil {
+				t.Log(err)
+			}
+			require.Equal(t, tt.desiredOutput, testResult)
+			if tt.shouldError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
```
nit: I would normally write it this way, I don't think there's a need to log expected errors or to check the result if there was an error
```suggestion
			if tt.shouldError {
				require.Error(t, err)
				return
			}
			require.NoError(t, err)
			require.Equal(t, tt.desiredOutput, testResult)
```

nklaassen commented on lib/puttyhosts/puttyhosts_test.go:
```diff
@@ -305,3 +309,132 @@ func TestNaivelyValidateHostname(t *testing.T) {
+		},
+		{
+			input:         "*.example.com && port:22",
+			desiredOutput: []string(nil),
```
nit: normally we just omit these empty/nil values for error cases

nklaassen commented on lib/puttyhosts/puttyhosts.go:
```diff
@@ -238,3 +238,40 @@ func FormatHostCAPublicKeysForRegistry(hostCAPublicKeys map[string][]string, hos
+				return nil, trace.BadParameter("validity string for %v contains invalid splitter token %q at position %d, see %v", caName, token, count, docsURL)
+			}
+		} else {
+			// if the string contains any value which is not || or part of a hostname, return an error
+			if strings.ContainsAny(token, "()&!:") {
```
I think we can/should reject `||` here as well

webvictim replied:
Good catch, `|| ||` is caught by the non-odd error but `|| || ||` would have made it through before.

nklaassen commented on lib/puttyhosts/puttyhosts_test.go:
```diff
@@ -305,3 +309,132 @@ func TestNaivelyValidateHostname(t *testing.T) {
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.input, func(t *testing.T) {
```
Sorry it's tedious but could we give each of the cases a name? It will should up nicer in the `go test` output and it serves as a comment, on some of the cases it isn't completely obvious why they should be errors
```suggestion
		t.Run(tt.name, func(t *testing.T) {
```

nklaassen commented on lib/puttyhosts/puttyhosts.go:
```diff
@@ -238,3 +238,40 @@ func FormatHostCAPublicKeysForRegistry(hostCAPublicKeys map[string][]string, hos
+	if len(splitTokens)%2 != 1 {
+		return nil, trace.BadParameter("validity string for %v contains non-odd [%d] number of entries, see %v", caName, len(splitTokens), docsURL)
+	}
+	for count, token := range splitTokens {
```
nit: I'd prefer to name it `i` or `index` instead of `count`, just a bit more conventional and because it's a bit weird for a "count" to start at 0

nklaassen commented on tool/tsh/common/putty_config_windows.go:
```diff
@@ -180,16 +185,47 @@ func addHostCAPublicKey(registryHostCAStruct puttyhosts.HostCAPublicKeyForRegist
+	if len(matchHosts) > 0 {
+		log.Debugf("Found %v legacy MatchHosts value(s) in registry key %v, migrating to new Validity format and deleting", len(matchHosts), registryKeyName)
+		hostList = append(hostList, matchHosts...)
+		err := registry.DeleteValueFromRegistryKey(registryKey, "MatchHosts")
```
Do you think we should keep setting MatchHosts for any older software that does not recognize Validity yet?

If not, perhaps we should wait to delete it until after we have succesfully written Validity, in case there is any error

webvictim replied:
From the `tsh` side, `puttyconfig` was only released in 14.0.0 so my thought is that very few people will have used it yet. I've not heard of anyone actually doing so.

From the PuTTY side, only the first alpha versions (which I based my original PR on) used the `MatchHosts` multi-string. Support for `Validity` was added somewhere during the beta testing and was the default/recommended approach on the release of 0.78, which was the first release that supported SSH certs. Truth be told it was laziness on my part and reluctance to rewrite/retest the PR which meant that this wasn't done until now - if I'd been more diligent, we could've/should've supported `Validity` on the initial release of `tsh puttyconfig` in 14.0.0.

This logic is essentially my effort to expunge `MatchHosts` from existence before it gets popular. If you feel strongly about it, maybe I could commit the logic commented out with a note/reminder to uncomment it in 14.1.0 or something? Alternatively, we could log any failure to delete at INFO level rather than DEBUG so people see it and raise bug reports, then move it back to DEBUG in future when we're more confident?

nklaassen replied:
Oh I didn't realize this was all so new, seems fine to delete the `MatchHosts` key then.

Don't comment it out, I meant that maybe we should move the deletion of the `MatchHosts` key to after the line where we write the `Validity` key, just to avoid potentially breaking a working setup and then failing to write the `Validity` key for some reason.

webvictim replied:
Done in https://github.com/gravitational/teleport/pull/32598/commits/257b4c001ea540f11935d375999359b72f4d07e8

webvictim committed:
Apply fixes from code review

webvictim committed:
Tidy up errors, provide consistent detail about which field contains an error

webvictim committed:
Disable docs lint for dots in heading

This is needed here, as there are 5 error messages which all start the same way but end differently.

webvictim committed:
Catch a few more error cases

webvictim committed:
Only delete old MatchHosts key after new Validity key has been written successfully

webvictim committed:
Merge branch 'master' into gus/tsh-putty-matchhosts-validity

zmb3 commented on tool/tsh/common/putty_config_windows.go:
```diff
@@ -207,6 +237,16 @@ func addHostCAPublicKey(registryHostCAStruct puttyhosts.HostCAPublicKeyForRegist
+	// if matchHosts has any entries, delete the "MatchHosts" key from the registry as its entries were migrated above.
+	if len(matchHosts) > 0 {
+		log.Debugf("Deleting %v legacy MatchHosts value(s) from registry key %v", len(matchHosts), registryKeyName)
+		err := registry.DeleteValueFromRegistryKey(registryKey, "MatchHosts")
```
All that `DeleteValueFromRegistryKey` does is call `registryKey.DeleteValue()` and log if there's an error. This code already logs if there's an error, so I don't see what we gain by adding an extra func.

Can we eliminate the duplicate logging and get rid of an exported func by removing `DeleteValueFromRegistryKey` here?

webvictim committed:
Apply suggestions from code review

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

webvictim committed:
Address Zac's comments from code review

webvictim committed:
Merge branch 'gus/tsh-putty-matchhosts-validity' of https://github.com/gravitational/teleport into gus/tsh-putty-matchhosts-validity

webvictim committed:
Merge branch 'master' into gus/tsh-putty-matchhosts-validity

webvictim committed:
Merge branch 'master' into gus/tsh-putty-matchhosts-validity
